1997 create update submission use case

erezmus opened the pull request:
#### Background

Creates an update submission use case to replace current update manuscript one. It does so by using the new Submission aggregate which as part of this PR has the update functionality added.

#### Any relevant tickets

Closes #1997.

#### How has this been tested?
**Reviewers** ensure that these test requirements are also reviewed.

Please indicate the need for this change to be tested. If not please justify, if so then at what levels. 

- [ ] Unit / Integration tests
- [ ] Browser tests

erezmus committed:
Part of #1997:
- Added functionality to Submission aggregate for submission update
- Added updateSubmissionUseCase
- Updated and refactored Submission unit tests

will-byrne committed:
add tests for teams

erezmus committed:
Part of #1997:
- Updated methods for updating teams for Submission aggregate and updated tests
- Started adding tests for updateSubmissionUseCase

erezmus committed:
Part of #1997: removed hasMatchingEditors and removed console log

erezmus committed:
Part of #1997: fixed issues in mergeObjects and resolver and aggregate tests

erezmus committed:
Part of #1997: fixed plurality for roles and added test for teams at updateManuscript resolver test

erezmus commented:
Still to do/think about:
- [ ] Move the updateManuscript.test.js from the resolvers into integration tests location ?
- [x] Remove the old updateManuscript.js file
- [x] Remove the Manuscript use case's `update` method

There's probably more.

erezmus committed:
Part of #1997:
- changed Manuscript model's save method back to using noDelete: true option for upsertGraphAndFetch
- added test for updateSubmissionUseCase to return a json object
- removed resolvers/updateManuscript.js
- update updateSubmissionUseCase to catch integrity errors

erezmus committed:
Merge branch 'develop' into 1997-create-update-submission-use-case

will-byrne commented on packages/component-submission/server/aggregates/Submission.test.js:
```diff
@@ -11,88 +13,92 @@ const createMockObject = (values = {}, mockSaveFn) => ({
+  beforeAll(() => {
+    mockManuscriptFind = jest.spyOn(Manuscript, 'find')
+    mockFileFind = jest.spyOn(File, 'findByManuscriptId')
+    mockTeamFind = jest.spyOn(Team, 'findByManuscriptId')
```
these three mocks can be instantiated outside the beforeAll() where they are declared.

erezmus committed:
Part of #1997: only load teams relation and disable AutoSave on ComponentWillUnmount hook as we are already saving on SavePage

will-byrne committed:
remove no longer necessary AutoSave Test

Cooryd commented on packages/component-submission/server/use-cases/updateSubmissionUseCase.js:
```diff
@@ -0,0 +1,60 @@
+      if (error.message.startsWith(expected)) {
+        logger.error(`Expected a ${expected}, ${error.message}`)
+      } else {
+        // not an error we were expecting.
```
Can probably remove this comment now https://github.com/elifesciences/elife-xpub/pull/2009/files#diff-5b518139d4766e3a75dc9c2cb1213c90R86 and https://github.com/elifesciences/elife-xpub/pull/2009/files#diff-5b518139d4766e3a75dc9c2cb1213c90R122 are expected error cases given 'bad' `data` objects.

Cooryd commented on packages/component-submission/server/aggregates/Submission.test.js:
```diff
@@ -192,33 +176,287 @@ describe('Submission', () => {
+    })
+  })
+
+  describe('updateEditorTeams', () => {
```
Can we add a test for the validation check here https://github.com/elifesciences/elife-xpub/pull/2009/files#diff-5b518139d4766e3a75dc9c2cb1213c90R122

Cooryd commented on packages/component-submission/server/aggregates/Submission.test.js:
```diff
@@ -192,33 +176,287 @@ describe('Submission', () => {
       )
```
I think we should have a unit test under the `updateManuscript` describe to check the status validation here: https://github.com/elifesciences/elife-xpub/pull/2009/files#diff-5b518139d4766e3a75dc9c2cb1213c90R84

petereast commented on packages/component-model-team/entities/team/index.js:
```diff
@@ -24,6 +24,18 @@ class Team extends BaseModel {
   }
+
+  static async findByManuscriptId(manuscriptId) {
+    const teams = await this.query().where({
```
Wait - static functions shouldn't have access to `this` - or am I getting confused?

petereast commented on packages/component-submission/server/use-cases/updateSubmissionUseCase.js:
```diff
@@ -0,0 +1,60 @@
+    } catch (error) {
+      const expected = 'Data Integrity Error'
+      if (error.message.startsWith(expected)) {
+        logger.error(`Expected a ${expected}, ${error.message}`)
```
Is it safe to just swallow these errors - is there anything we need to do to clean up after the error occurs? Also which of the above statements throws - if it throws will it leave the stored data in an error state or will the transaction be aborted?

petereast replied:
Upon further inspection, if one of the `submission.update*` fails, the there's potential for a partial object to be stored and the error be swallowed resulting in :warning:  **loss of data**. :warning: 

I suggest a change is made to each of the `submission.update*` functions allowing the user to pass in a `transaction` object that's used by the internal database calls so these updates can be rolled back if one of them fails.

erezmus replied:
The problem is that internally it calls the Manuscript model's save method which has a transaction and throws if the record has a later timestamp. Not sure how nesting transactions work and rewriting the manuscript save method is too big a change. I do agree however that it should be all one transaction

petereast replied:
Maybe we could just delete everything related to this `Manuscript` if an error occurs? The way it is currently, if an error happens at any stage we're going to end up with partial/corrupted data in the database, I can't think of anything else we could do without introducing transactions at this level.

erezmus replied:
I think some of the teams not being saved is acceptable for now and that would be reflected on the client once the new data has been fetched.

erezmus committed:
Part of #1997:
- Added tests for Submission aggregate to check status and editor conflict assertions and moved setting of mocks
- Improved logging message for savePage mutation
